assayist/client/modify.py

```diff
@@ -0,0 +1,104 @@
+    # alternative name with the passed in canonical name and alternatives
+    names_where_clause = [
+        'c.canonical_name = "{0}" OR "{0}" in c.alternative_names'.format(name)
+        for name in itertools.chain([c_name], alternatives)
+    ]
+    query = """
+    MATCH (c:Component {{canonical_type: "{}", canonical_namespace: "{}" }})
+    WHERE {}
+    RETURN c
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
+
+    # If no matching component is returned, just create one
+    if not components:
+        component = Component(canonical_namespace=c_namespace, canonical_name=c_name,
+                              canonical_type=c_type, alternative_names=alternatives).save()
+        log.info(f'Creating the component "{component}"')
+        return
+
+    # This will be the only remaining component if there is more than one component returned, as
+    # the information stored in the others will be merged into this one
+    component = components[0]
+    log.info(f'Modifying the existing component "{component}"')
+    if component.canonical_name != c_name:
+        # Add the old canonical name as an alternative name
+        if component.canonical_name not in component.alternative_names:
+            component.alternative_names.append(component.canonical_name)
+        # Remove the new canonical name from the alternative names if it's present
+        if c_name in component.alternative_names:
+            component.alternative_names.pop(component.alternative_names.index(c_name))
+        # Set the proper canonical name
+        component.canonical_name = c_name
+
+    # Add the missing alternative names
+    for alternative in alternatives:
+        if alternative not in component.alternative_names:
+            component.alternative_names.append(alternative)
+
+    # If there was more than one component that matched, we must merge them
+    for c in components[1:]:
+        log.info(f'Merging the component "{c}" in the existing component "{component}"')
+        # Merge all the source locations on the first component
+        for sl in c.source_locations.all():
+            component.source_locations.connect(sl)
+
+        # Add the canonical name of this component as an alternative name for the first component,
+        # if applicable
+        if c.canonical_name != component.canonical_name and \
+                c.canonical_name not in component.alternative_names:
+            component.alternative_names.append(c.canonical_name)
+        # Add all the alternative names of this component to the first component, if applicable
+        for alternative in c.alternative_names:
+            if alternative != component.canonical_name and \
+                    alternative not in component.alternative_names:
+                component.alternative_names.append(alternative)
```

Comment:
All of the altnerative_names loops and if statements in the last 40ish lines can probably be more simply handled by:
1) Emptying component.alternative_names
2) Adding the desired entries to a set
3) Iteratively calling component.alternative_names.append at the very end.